docs/abi.rst

```diff
@@ -0,0 +1,560 @@
+.. _abi:
+
+ABI Support
+===========
+
+TODO: brief description
+
+.. warning::
+  ABI support is still taking shape and is subject to backwards incompatible changes.
+    * Based on feedback, the API and usage patterns are likely to change.
+    * For the following use cases, feel encouraged to rely on abstractions.  Expect a best-effort attempt to minimize backwards incompatible changes along with a migration path.
+
+       * :any:`ABIReturnSubroutine` usage for ABI Application entry point definition.
+
+       * :any:`Router` usage for defining how to route program invocations.
+    * For general purpose :any:`Subroutine` definition usage, use at your own risk.  Based on feedback, the API and usage patterns will change more freely and with less effort to provide migration paths.
+
+TODO: warning about how ABI types are intended to be used at the moment (for standard serialization at the incoming and outgoing layers of contract) - internal use may be extremely inefficient with ABI types
+TODO: warning about how internal representation of ABI types may change over time
+
+Types
+------
+
+The ABI supports a variety of types whose encodings are standardized.
+
+Fundamentals
+~~~~~~~~~~~~
+
+Before we introduce these ABI types, it's important to understand the fundamentals of how PyTeal's ABI type system works.
+
+:code:`abi.BaseType`
+^^^^^^^^^^^^^^^^^^^^
+
+:any:`abi.BaseType` is an abstract base class that all ABI type classes inherit from. This class defines a few methods common to all ABI types:
+
+* :any:`abi.BaseType.decode()` is used to decode and populate a type's value from an encoded byte string.
+* :any:`abi.BaseType.encode()` is used to encode a type's value into an encoded byte string.
+* :any:`abi.BaseType.type_spec()` is used to get an instance of :any:`abi.TypeSpec` that describes that type.
+
+:code:`abi.TypeSpec`
+^^^^^^^^^^^^^^^^^^^^
+
+:any:`abi.TypeSpec` is an abstract base class used to describe ABI types. Every child class of :code:`abi.BaseType` also has a companion :code:`abi.TypeSpec` child class. The :code:`abi.TypeSpec` class has a few methods that return information about the type it represents, but one of the class's most important features is the method :any:`abi.TypeSpec.new_instance()`, which creates and returns a new :any:`abi.BaseType` instance of the ABI type it represents.
+
+Static vs Dynamic Types
+^^^^^^^^^^^^^^^^^^^^^^^
+
+An important property of an ABI type is whether it is static or dynamic.
+
+Static types are defined as types whose encoded length does not depend on the value of that type. This property allows encoding and decoding of static types to be more efficient.
```

Comment:
perhaps you could add "(for example: a boolean)" and "(for example: a string)" to exemplify what is meant in the section.